shared/shared/postgresql_backend/sql_read_database_backend_service.py

```diff
@@ -48,51 +279,123 @@ def delete_models(self, fqids: List[str]) -> None:
         # and delete_fields events.
+        # Optionally only builds the models up to the specified position.
         # TODO: There might be a speedup: Get the model from the readdb first.
         # If the model exists there, read the position from it, use the model
         # as a starting point in `build_model_from_events` and just fetch all
         # events after the position.
+        if position:
+            pos_cond = "and position <= %s"
+            pos_args = [position]
+        else:
+            pos_cond = ""
+            pos_args = []
+
         included_types = dedent(
             f"""\
             ('{EVENT_TYPES.CREATE}',
             '{EVENT_TYPES.UPDATE}',
             '{EVENT_TYPES.DELETE_FIELDS}')"""
         )
-        query = (
-            "select type, data from events where fqid=%s "
-            + f"and type in {included_types} order by id asc"
+        query = dedent(
+            f"""\
+            select fqid, type, data, position from events e
+            where fqid in %s and type in {included_types} {pos_cond}
+            order by id asc"""
         )
 
-        events = self.connection.query(query, [fqid])
-        return self.build_model_from_events(events)
+        args: List[Any] = [tuple(fqids)]
+        db_events = self.connection.query(query, args + pos_args)
 
-    def build_model_from_events(
-        self, events: List[Tuple[int, Dict[str, Any]]]
-    ) -> Dict[str, Any]:
+        partitions = {}
+        for event in db_events:
+            if event["fqid"] not in partitions:
+                partitions[event["fqid"]] = [event]
+            else:
+                partitions[event["fqid"]] += [event]
+
+        models = {}
+        for fqid, events in partitions.items():
+            models[fqid] = self.build_model_from_events(events)
+
+        return models
+
+    def build_model_from_events(self, events: List[Dict[str, Any]]) -> Model:
         if not events:
             raise BadCodingError()
 
         create_event = events[0]
-        assert create_event[0] == EVENT_TYPES.CREATE
-        model = create_event[1]
+        assert create_event["type"] == EVENT_TYPES.CREATE
+        model = create_event["data"]
 
         # apply all other update/delete_fields
         for event in events[1:]:
-            if event[0] == EVENT_TYPES.UPDATE:
-                model.update(event[1])
-            elif event[0] == EVENT_TYPES.DELETE_FIELDS:
-                for field in event[1]:
+            if event["type"] == EVENT_TYPES.UPDATE:
+                model.update(event["data"])
+            elif event["type"] == EVENT_TYPES.DELETE_FIELDS:
+                for field in event["data"]:
                     if field in model:
                         del model[field]
             else:
                 raise BadCodingError()
 
+        model[META_POSITION] = events[-1]["position"]
         return model
 
+    def is_deleted(self, fqid: str, position: Optional[int] = None) -> bool:
```

Comment:
Note: This may be removed, when applying the proposed changes.

Reply:
Is still needed if `get` is called with a position. Could remove the optionality of the parameter, but would rather keep it if it is ever needed, since the tests for it already exist too.